### PR TITLE
chore: bump network controller 20.0.0

### LINKED ISSUE #11156: Migrate Network Controller from Version 19 to Version 20
### What is this about?

We need to migrate the Network Controller from version 19 to version 20. This migration involves updating the state structure to align with the new version's requirements. The key changes include the removal of the providerConfig object and the addition of new network configurations. This update ensures compatibility with the latest features and improvements in Network Controller version 20.

### Scenario

_No response_

### Design

_No response_

### Technical Details

_No response_

### Threat Modeling Framework

_No response_

### Acceptance Criteria

_No response_

### Stakeholder review needed before the work gets merged

- [X] Engineering (needed in most cases)
- [ ] Design
- [ ] Product
- [ ] QA (automation tests are required to pass before merging PRs but not all changes are covered by automation tests - please review if QA is needed beyond automation tests)
- [ ] Security
- [ ] Legal
- [ ] Marketing
- [ ] Management (please specify)
- [ ] Other (please specify)

### References

_No response_

### PR DESCRIPTION
**Pull request by salimtb**
## **Description**

This PR migrates the Network Controller from version 19 to version 20. The migration involves updating the state structure to align with the new version's requirements. The key changes include the removal of the providerConfig object and the addition of new network configurations. This update ensures compatibility with the latest features and improvements in Network Controller version 20.

## **Related issues**

Fixes: #11156 

## **Manual testing steps**

1. Verify that the providerConfig data has been moved to the networkConfigurations object.
2. Check that the selectedNetworkClientId is updated to the new network configuration ID.
3. Ensure that all existing network configurations and metadata are preserved and correctly formatted.
4. Perform various network operations to ensure the application functions correctly with the new state structure.

### **After**

https://drive.google.com/drive/folders/1AyR4RNSlySfy1X64UwL-gZ7mowN0whOP?usp=drive_link

## **Pre-merge author checklist**

- [x] I’ve followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Mobile Coding Standards](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've completed the PR template to the best of my ability
- [x] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [x] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

### CODE REVIEW

File: app/components/Views/confirmations/components/TransactionReview/TransactionReviewData/index.test.tsx [hunk 1]
```diff
@@ -6,10 +6,9 @@ import { Provider } from 'react-redux';
 import { backgroundState } from '../../../../../../util/test/initial-root-state';
 
 const mockStore = configureMockStore();
+
 const initialState = {
-  engine: {
-    backgroundState,
-  },
+  backgroundState,
```
Comment (tommasini) [inline app/components/Views/confirmations/components/TransactionReview/TransactionReviewData/index.test.tsx:11]: This doesn't seem right, backgroundState is inside of the "engine" reducer, and it's where the component finds the data with this [selectors](https://github.com/MetaMask/metamask-mobile/blob/57b556e872dcc4101ed7d29cb4fca22bf948d462/app/components/Views/confirmations/components/TransactionReview/TransactionReviewData/index.js#L186) 

I think we still would need this data here, I think it was not flagged by unit tests on the snapshot because this component still uses enzyme for create the snapshot

File: app/store/migrations/034.ts [hunk 2]
```diff
@@ -1,6 +1,17 @@
 import { captureException } from '@sentry/react-native';
 import { isObject, hasProperty } from '@metamask/utils';
-import type { NetworkState } from '@metamask/network-controller';
+
+interface NetworkState {
```
Comment (tommasini) [inline app/store/migrations/034.ts:4]: Thanks for fixing this! The future appreciates it!

File: app/util/test/initial-background-state.json [hunk 3]
```diff
@@ -58,11 +58,6 @@
   },
   "NetworkController": {
     "networkConfigurations": {},
-    "providerConfig": {
```
Comment (tommasini) [inline app/util/test/initial-background-state.json:61]: Shouldn't we fill this with the default state of the NetworkController? I'm checking on main, but mainnet is still the default network: https://github.com/MetaMask/core/blob/442a30075434b07c995cf09f3831041e93046b3d/packages/network-controller/src/NetworkController.ts#L509

File: app/selectors/networkController.ts [hunk 4]
```diff
@@ -1,15 +1,54 @@
...
 export const selectProviderConfig = createDeepEqualSelector(
   selectNetworkControllerState,
-  (networkControllerState: NetworkState) =>
-    networkControllerState?.providerConfig,
+  (networkControllerState: NetworkState) => {
+    const { NetworkController } = Engine?.context || {};
+
+    const builtInNetwork =
+      // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+      // @ts-ignore
```
Comment (tommasini) [inline app/selectors/networkController.ts:28]: I believe as a last resource if Typescript isn't catching the types we could cast it, but let's avoid this ts-ignore 
Please check out: https://github.com/MetaMask/contributor-docs/blob/main/docs/typescript.md#type-assertions

**Comment by tommasini** [comment]
@salimtb The PR looks good, I wanted to ask you if you could record some scenarios of network interactions as you mentioned here on the manual testing steps:
> Perform various network operations to ensure the application functions correctly with the new state structure.

* Add a network from popular list, switch to that network, kill the app, open it, that network is still selected
* Add custom network, switch to that network, kill the app, open it, that network is still selected
* Add network via dapp, switch to that network,  interact with the dapp, ensure that it's that the network being used
* Go to settings, network settings, edit an network details, kill the app, open it, ensure that the details are the new ones
* Delete a network
* Performa a transaction with a layer 1 and, layer 2 and a test network

File: app/selectors/networkController.ts (hunk 4 above)
Comment (tommasini) [reply app/selectors/networkController.ts:28]: I'm super interested on improve my typescript best pratices as well! If you want feel free to ping me for we to see the best solution here!

File: app/store/migrations/053.test.ts [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+      },
+    };
+
+    const expectedState = {
+      engine: {
+        backgroundState: {
+          NetworkController: {
+            selectedNetworkClientId: 'mainnet',
+            networksMetadata: {},
+            networkConfigurations: {},
```
Comment (tommasini) [inline app/store/migrations/053.test.ts:95]: Cool, we do expect that configurations be empty at this point since we are leaving for the NetworkController populate this values once we initialise it [here](https://github.com/MetaMask/core/blob/442a30075434b07c995cf09f3831041e93046b3d/packages/network-controller/src/NetworkController.ts#L857) correct?

File: patches/@metamask+ppom-validator+0.32.0.patch [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+         // register message handlers
+@@ -177,6 +188,20 @@ class PPOMController extends base_controller_1.BaseControllerV2 {
+     }
+ }
+ exports.PPOMController = PPOMController;
++const getCurrentNetwork = (networkControllerState) => {
++      const configurations = Object.values(networkControllerState.networkConfigurations);
++      if (configurations.length === 0) {
++          console.warn("No network configurations available.");
++          // Handle the empty case, maybe return a default value or halt further execution.
```
Comment (tommasini) [inline patches/@metamask+ppom-validator+0.32.0.patch:47]: This have a comment that most likely is saying that it's still is work in progress, just checking if it is

File: patches/@metamask+ppom-validator+0.32.0.patch [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
++      } else {
++          const currentNetwork = getCurrentNetwork(networkControllerState);
++          if (currentNetwork && currentNetwork.chainId) {
++              // Proceed with using chainId
```
Comment (tommasini) [inline patches/@metamask+ppom-validator+0.32.0.patch:51]: Here as well

File: app/constants/network.js [hunk 8]
```diff
@@ -7,7 +7,7 @@ export const SEPOLIA = 'sepolia';
 export const LINEA_GOERLI = 'linea-goerli';
 export const LINEA_SEPOLIA = 'linea-sepolia';
 export const LINEA_MAINNET = 'linea-mainnet';
-export const RPC = NetworkType.rpc;
+export const RPC = 'rpc';
```
Comment (bergeron) [inline app/constants/network.js:10]: `NetworkType.rpc` should still exist in this version of `controller-utils`, if you want to try to revert this change

File: app/util/test/network.ts [hunk 9]
```diff
@@ -0,0 +1,60 @@
+// TODO: This is intentionally the old network state, and could be
+
+import { NetworkMetadata, NetworkStatus } from '@metamask/network-controller';
+import { Hex } from '@metamask/utils';
+import { v4 as uuidv4 } from 'uuid';
+
+// removed if the e2e tests bump `FIXTURE_STATE_METADATA_VERSION` to latest
+// eslint-disable-next-line import/prefer-default-export
+export const mockNetworkStateOld = (
```
Comment (bergeron) [inline app/util/test/network.ts:9]: I think this comment only applies on extension.  Also it probably doesn't need to be suffixed with `Old`.  That was some extension specific behavior where we have to mock it in both the old and new ways, because it runs the e2e test state through migrations.

File: app/util/test/initial-background-state.json (hunk 3 above)
Comment (salimtb) [reply app/util/test/initial-background-state.json:61]: on your link you're refering to the version 21 , the correct one is the function `getNetworkConfigurationByNetworkClientId` , mainnet is populated by default from the [BUILT_IN_NETWORKS](https://github.com/MetaMask/core/blob/29e75c39dd3679e58af69671c7df5f1094f07488/packages/controller-utils/src/constants.ts#L56)

File: app/store/migrations/053.test.ts (hunk 5 above)
Comment (salimtb) [reply app/store/migrations/053.test.ts:95]: correct yes, default network will be populated on the initiallisation

File: app/store/migrations/053.ts [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+
+  if (!isObject(state.engine.backgroundState)) {
+    captureException(
+      new Error(
+        `FATAL ERROR: Migration 53: Invalid engine backgroundState error: '${typeof state
+          .engine.backgroundState}'`,
+      ),
+    );
+    return state;
+  }
```
[the remark this answers is not in the document]
Comment (salimtb) [reply app/store/migrations/053.ts:28]: not sure i get what you mean here

**Comment by salimtb** [comment]
> @salimtb The PR looks good, I wanted to ask you if you could record some scenarios of network interactions as you mentioned here on the manual testing steps:
> 
> > Perform various network operations to ensure the application functions correctly with the new state structure.
> 
> * Add a network from popular list, switch to that network, kill the app, open it, that network is still selected
> * Add custom network, switch to that network, kill the app, open it, that network is still selected
> * Add network via dapp, switch to that network,  interact with the dapp, ensure that it's that the network being used
> * Go to settings, network settings, edit an network details, kill the app, open it, ensure that the details are the new ones
> * Delete a network
> * Performa a transaction with a layer 1 and, layer 2 and a test network

hey @tommasini , [here](https://drive.google.com/drive/folders/1AyR4RNSlySfy1X64UwL-gZ7mowN0whOP?usp=sharing) is the list of tests you requested

**Review by tommasini — DISMISSED** [review]

File: app/components/UI/AccountInfoCard/index.test.tsx [hunk 11]
```diff
@@ -48,11 +48,19 @@ const mockInitialState: DeepPartial<RootState> = {
...
-          type: 'sepolia',
-          nickname: 'Sepolia',
-          ticker: 'ETH',
+        selectedNetworkClientId: 'sepolia',
+        networksMetadata: {},
+        networkConfigurations: {
+          sepolia: {
+            id: 'sepolia',
+            rpcUrl: 'http://localhost/v3/',
+            chainId: '0xaa36a7',
```
Comment (gambinish) [inline app/components/UI/AccountInfoCard/index.test.tsx:57]: Should this use `CHAIN_ID`?

File: app/components/Views/confirmations/components/TransactionReview/__snapshots__/index.test.tsx.snap [hunk 12]
```diff
@@ -402,7 +404,7 @@ exports[`TransactionReview should match snapshot 1`] = `
                                       }
                                     }
                                   >
-                                    Sepolia
+                                    Ethereum Main Network
```
Comment (gambinish) [inline app/components/Views/confirmations/components/TransactionReview/__snapshots__/index.test.tsx.snap:407]: Is this diff expected?

File: app/components/hooks/useBlockExplorer.test.ts [hunk 13]
```diff
@@ -12,10 +11,25 @@ const mockInitialState = {
     backgroundState: {
       ...backgroundState,
       NetworkController: {
-        providerConfig: {
-          type: LINEA_GOERLI,
-          rpcUrl: 'https://mainnet.infura.io/v3/1234567890abcdef',
+        networkConfigurations: {
+          linea_goerli: {
+            chainId: '0xe704',
```
Comment (gambinish) [inline app/components/hooks/useBlockExplorer.test.ts:16]: I also wonder if this should use `CHAIN_IDS.GOERLI`

File: app/components/Views/confirmations/components/TransactionReview/__snapshots__/index.test.tsx.snap (hunk 12 above)
Comment (salimtb) [reply app/components/Views/confirmations/components/TransactionReview/__snapshots__/index.test.tsx.snap:407]: yes this diff is expected because we're changing the used network on the state

**Review by tommasini — APPROVED** [review]

File: app/components/UI/AccountInfoCard/index.test.tsx (hunk 11 above)
Comment (salimtb) [reply app/components/UI/AccountInfoCard/index.test.tsx:57]: this will be addressed on the v21 PR
